specification/sections/06-tracking.md

```diff
@@ -0,0 +1,40 @@
+---
+title: Tracking
+description: Specification defining tracking API
+toc_max_heading_level: 4
+---
+
+# 6. Tracking
+
+[![experimental](https://img.shields.io/static/v1?label=Status&message=experimental&color=orange)](https://github.com/open-feature/spec/tree/main/specification#experimental)
+
+## Overview
+
+Experimentation is a primary use case for feature flags.
+In practice, this often means flag variants are assigned to users at random or in accordance with a business rule, while the impact of the assigned variant on some business objective is measured.
+Vendors and custom solutions often support a _tracking_ or _goal measuring_ API to facilitate the measurement of these business objectives.
+
+### Goals
+
+- Develop official terminology to support consistent implementation
+- Specify a flexible API widely compatible with basic vendor functionality
+  - Define tracking event payload
+  - Define tracking event identifier
+  - Support client and server paradigms
```

Comment:
Should this include recommendations and/or SDK support for asynchronous events flushing? Most vendors support async events.

Reply:
You mean able to trigger event flushing manually, alongside it being part of the `shutdown`? That make sense I think

Reply:
That, but also how to approach async in general, perhaps the SDKs could bundle a robust implementation of message batching so that individual providers are easier to write.

Reply:
My sense is that SDKs and framework integrations will have a better context to handle these than individual providers.

Reply:
async can come together with batch aggregations to reduce network requests and possibly improve performance for high rate, and also via a hook, like done in [Go Feature Flag provider](https://github.com/open-feature/java-sdk-contrib/blob/main/providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/hook/DataCollectorHook.java).
it was described at the requirements of this example by @thomaspoignant:

>
- The data collection should be activated by default, but it should be possible to deactivate it in the configuration.
- We should not call the ... API ... for every evaluation, but call it in bulk with a periodic threshold.
- The threshold should be configurable by the user of the provider.

All these are related to implementation, so not sure whether or how to include mentioning it in spec.

Reply:
Agree, just pointing out that these choices impact the design of the provider API.

Reply:
\* fixing user

async can come together with batch aggregations to reduce network requests and possibly improve performance for high rate, and also via a hook, like done in [Go Feature Flag provider](https://github.com/open-feature/java-sdk-contrib/blob/main/providers/go-feature-flag/src/main/java/dev/openfeature/contrib/providers/gofeatureflag/hook/DataCollectorHook.java).
it was described at the requirements of this example by @thomaspoignant:

>
- The data collection should be activated by default, but it should be possible to deactivate it in the configuration.
- We should not call the ... API ... for every evaluation, but call it in bulk with a periodic threshold.
- The threshold should be configurable by the user of the provider.

All these are related to implementation, so not sure whether or how to include mentioning it in spec.

> Agree, just pointing out that these choices impact the design of the provider API and thus the spec

Reply:
As an end user of an event API I would expect the API from OF for tracking an event to be just a fire and forget. But I don't know how deep we want to formulate that here?

Reply:
> As an end user of an event API I would expect the API from OF for tracking an event to be just a fire and forget. But I don't know how deep we want to formulate that here?

Yes I agree.

I think @federicobond has a good point in that caching and flushing will be important to implementations, but I think that from an OpenFeature _application author_ perspective, this would be transparent. There may be some implications on defining some kind of lifecycle or connecting a hypothetical tracking provider with the existing feature provider lifecycle, to flush data on shutdown, etc.

Reply:
I will add a point about defining a lifecycle to support existing solutions (caching, flushing, batching, etc)... this also might just meaning hooking into the existing flag provider lifecycle which has similar needs.

@nicklasl @federicobond @liran2000 does that makes sense to you guys?

Reply:
Fine by me.
What about where to implement it - on the provider itself, or as a hook?

Reply:
Agree, from an _application author_ perspective it should be transparent. I was looking at it from a _provider author_ perspective.

Reply:
I think it is a reasonable goal to have a _decision_ on where the work of things like message batching, flushing and caching takes place.

Reply:
Great question, but I think it's a technical one for later.

Reply:
I think what you said @federicobond makes sense and will be needed but this feels like being part of 
`Specify a flexible API widely compatible with basic vendor functionality`.

`I think it is a reasonable goal to have a decision on where the work of things like message batching, flushing and caching takes place.`

Think so too @nicklasl, but maybe there is a more high level goal we can define for this.
But I am struggling to come up with a more general "subgoal".

Reply:
I've added `Support underlying SDK technical requirements (batching, caching, and flushing)`. I have intentionally kept this vague since this is a goals document.